Sources/llbuild2/Core/Engine.swift

```diff
@@ -196,7 +196,7 @@ extension Key: Hashable {
 
 public class LLBEngine {
     private let group: LLBFuturesDispatchGroup
-    private let delegate: LLBEngineDelegate
+    private weak var delegate: LLBEngineDelegate!
```

Comment:
I'm not really sure if the IUO is appropriate here, but it does seem like programmer error if the engine outlives the delegate.

Reply:
this is a breaking change, though, right?

Reply:
Yep.  As visible in the rest of the diff, for example, we were relying on the engine to keep the delegate alive.

Reply:
Also, the delegate now needs to be a `class`, whereas, again, like in the rest of the diff, we had a `struct` implementation.

Reply:
Maybe "delegate" is really the wrong shape to be using here?  If we'd take a block instead, for example, the caller would've been responsible for managing the memory semantics.

Reply:
Regarding the block/closure: Hmm, that however leaves you with just one callback not multiple like a delegate allows...

Reply:
What about making the delegate `strong` but passing the engine into every single delegate callback? Then the delegate has no reason to retain the engine?

Reply:
Yeah, that's kind of what I'm getting at.  It's weird no matter how you cut it, i.e. we have to answer uncomfortable questions about what to do in a case that we sort of hope never happens.  It'd be way better if we had a straightforward structure that made the expectations clear.

I _think_ either `shutdown()` or taking a block could do that.

Reply:
`shutdown` is always a good idea IMHO (and `assert` that it has been done in `deinit`). The whole Swift on Server ecosystem doesn't have `weak`s and stuff works just fine :).